numpyro/distributions/directional.py

```diff
@@ -19,6 +19,149 @@
+        must be less than or equal to one. See eq. 2.1 in [1].
+
+    ** References: **
+        1. Sine-skewed toroidal distributions and their application in protein bioinformatics
+            Ameijeiras-Alonso, J., Ley, C. (2019)
+
+    :param torch.distributions.Distribution base_dist: base density on a d-dimensional torus. Supported base
+        distributions include: 1D :class:`~numpyro.distributions.VonMises`,
+        :class:`~numnumpyro.distributions.SineBivariateVonMises`, 1D :class:`~numpyro.distributions.ProjectedNormal`,
+        and :class:`~numpyro.distributions.Uniform` (-pi, pi).
+    :param torch.tensor skewness: skewness of the distribution.
+    """
+
+    arg_constraints = {
+        "skewness": constraints.independent(constraints.interval(-1.0, 1.0), 1)
+    }
+
+    support = constraints.independent(
+        constraints.real, 1
+    )  # TODO: add Circular constraint (issue 1070 and PR 1080)
+
+    def __init__(self, base_dist: Distribution, skewness, validate_args=None):
+        assert (
+            base_dist.event_shape == skewness.shape[-1:]
+        ), "Sine Skewing is only valid with a skewness parameter for each dimension of `base_dist.event_shape`."
+
+        batch_shape = jnp.broadcast_shapes(base_dist.batch_shape, skewness.shape[:-1])
+        event_shape = skewness.shape[-1:]
+        self.skewness = jnp.broadcast_to(skewness, batch_shape + event_shape)
+        self.base_dist = base_dist.expand(batch_shape)
+        super().__init__(batch_shape, event_shape, validate_args=validate_args)
+
+    def __repr__(self):
+        args_string = ", ".join(
+            [
+                "{}: {}".format(
+                    p,
+                    getattr(self, p)
+                    if getattr(self, p).numel() == 1
+                    else getattr(self, p).size(),
+                )
+                for p in self.arg_constraints.keys()
+            ]
+        )
+        return (
+            self.__class__.__name__
+            + "("
+            + f"base_density: {str(self.base_dist)}, "
+            + args_string
+            + ")"
+        )
+
+    def sample(self, key, sample_shape=()):
+        base_key, skew_key = random.split(key)
+        bd = self.base_dist
+        ys = bd.sample(base_key, sample_shape)
+        u = random.uniform(skew_key, sample_shape + self.batch_shape)
+
+        # Section 2.3 step 3 in [1]
+        mask = u <= 0.5 + 0.5 * (
+            self.skewness * jnp.sin((ys - bd.mean) % (2 * jnp.pi))
+        ).sum(-1)
+        mask = mask[..., None]
+        samples = (jnp.where(mask, ys, -ys + 2 * bd.mean) + jnp.pi) % (
+            2 * jnp.pi
+        ) - jnp.pi
+        return samples
+
+    def log_prob(self, value):
+        if self._validate_args:
+            self._validate_sample(value)
+        if self.base_dist._validate_args:
+            self.base_dist._validate_sample(value)
+
+        # Eq. 2.1 in [1]
+        skew_prob = jnp.log(
```

Comment:
why not use `log1p` as in Pyro?

Reply:
I copied over a commit without all revisions. Sorry for making you review the same thing twice.